kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/ProxyChannelStateMachine.java

```diff
@@ -0,0 +1,524 @@
+     *
+     * @param msg the message to be <em>logged</em> in explanation of the error condition
+     */
+    void illegalState(@NonNull String msg) {
+        if (!(state instanceof Closed)) {
+            LOGGER.error("Unexpected event while in {} message: {}, closing channels with no client response.", state, msg);
+            toClosed(null);
+        }
+    }
+
+    /**
+     * A message has been received from the upstream node which should be passed to the downstream client
+     * @param msg the object received from the upstream
+     */
+    void forwardToClient(Object msg) {
+        Objects.requireNonNull(frontendHandler).forwardToClient(msg);
+    }
+
+    /**
+     * Called to notify the state machine that reading the upstream batch is complete.
+     */
+    void serverReadComplete() {
+        Objects.requireNonNull(frontendHandler).flushToClient();
+    }
+
+    /**
+     * A message has been received from the downstream client which should be passed to the upstream node
+     * @param msg the RPC received from the upstream
+     */
+    void forwardToServer(Object msg) {
+        Objects.requireNonNull(backendHandler).forwardToServer(msg);
+    }
+
+    /**
+     * Called to notify the state machine that reading the downstream the batch is complete.
+     */
+    void clientReadComplete() {
+        if (state instanceof Forwarding) {
+            Objects.requireNonNull(backendHandler).flushToServer();
+        }
+    }
+
+    /**
+     * The proxy has received something from the client. The current state of the session determines what happens to it.
+     * @param dp the decode predicate to be used if the session is still being negotiated
+     * @param msg the RPC received from the downstream client
+     */
+    void onClientRequest(@NonNull SaslDecodePredicate dp,
+                         Object msg) {
+        Objects.requireNonNull(frontendHandler);
+        if (state() instanceof Forwarding) { // post-backend connection
+            forwardToServer(msg);
+        }
+        else {
+            if (!onClientRequestBeforeForwarding(dp, msg)) {
+                illegalState("Unexpected message received: " + (msg == null ? "null" : "message class=" + msg.getClass()));
+            }
+        }
+    }
+
+    /**
+     * ensure the state machine is in the connecting state.
+     * @param msg to be logged if in another state.
+     */
+    void assertIsConnecting(String msg) {
+        if (!(state instanceof ProxyChannelState.Connecting)) {
+            illegalState(msg);
+        }
+    }
+
+    /**
+     * ensure the state machine is in the selecting server state.
+     *
+     * @return the SelectingServer state
+     * @throws IllegalStateException if the state is not {@link ProxyChannelState.SelectingServer}.
+     */
+    ProxyChannelState.SelectingServer enforceInSelectingServer(String errorMessage) {
```

Comment:
Every usage of this currently passes the same message string however I hesitate to inline it as the statemachine is not a collaborator of the `NetFilter` so it would be strange for it have a message mentioning that...